cmd/fpga_plugin/README.md

```diff
@@ -1,120 +1,269 @@
+    resource names, as advertised by the device plugin.
+
+    The webhook can also set environment variables to instruct the FPGA prestart hook to program the FPGA
+    before launching the container.
+
 -   [FPGA prestart CRI-O hook](../fpga_crihook/README.md)
 
+    A [CRI-O](https://github.com/cri-o/cri-o) prestart hook that, upon instruction from the FPGA admission
+    controller, allocates and programs the FPGA before the container is launched.
 
-### Get source code:
-```
-    $ mkdir -p $GOPATH/src/github.com/intel/
-    $ cd $GOPATH/src/github.com/intel/
-    $ git clone https://github.com/intel/intel-device-plugins-for-kubernetes.git
-```
+The repository also contains an [FPGA helper tool](../fpga_tool/README.md) that may be useful during
+development, initial deployment and debugging.
 
-### Build FPGA device plugin:
-```
-    $ cd $GOPATH/src/github.com/intel/intel-device-plugins-for-kubernetes
-    $ make fpga_plugin
-```
+# Installation
+
+The below sections cover how to obtain, build and install this component.
+
+Components can generally be installed either using DaemonSets or running them
+'by hand' on each node.
+
+## Pre-built images
+
+Pre-built images of the components are available on the [Docker hub](https://hub.docker.com/u/intel).
+These images are automatically built and uploaded from the latest `master` branch of this repository.
+
+Release tagged images of the components are also available on the Docker hub, tagged with their
+release version numbers.
+
+The deployment YAML files supplied with these components in this repository use the images with the
+`devel` tag by default. If you do not build your own local images, then your Kubernetes cluster may
+pull down the `devel` images from the Docker hub by default.
+
+To use the release tagged versions of the images, edit the YAML deployment files appropriately.
+
+The following images are available on the Docker hub:
+
+- [The FPGA plugin](https://hub.docker.com/r/intel/intel-fpga-plugin)
+- [The FPGA admisson webhook](https://hub.docker.com/r/intel/intel-fpga-admissionwebhook)
+- [The FPGA CRI-O plugin (in the `initcontainer` image)](https://hub.docker.com/r/intel/intel-fpga-initcontainer)
+
+## Dependencies
+
+All components have the same basic dependancies as the
+[generic plugin framework dependencies](../../README.md#about)
+
+To obtain a fully operational FPGA enabled cluster, you must install all three
+major components:
+
+-   [FPGA device plugin](README.md) (this component)
+-   [FPGA admission controller webhook](../fpga_admissionwebhook/README.md)
+-   [FPGA prestart CRI-O hook](../fpga_crihook/README.md)
+
+The CRI-O hook is only *required* if orchestrated FPGA bitstream programming (mode) is
+being used, but is installed by default by the
+[FPGA plugin DaemonSet YAML](../../deployments/fpga_plugin/fpga_plugin.yaml). If not using
+the FPGA admission controller webhook orchestration mode, and therefore not using the
```

Comment:
This should be refrased I think. orchestration mode is not only used for webhook. FPGA plugin must run in the same mode as webhook for the whole workflow to work. In orchestration programmed or 'region' mode plugin announces FPGA regions to kubelet. In preprogrammed or 'af' mode it announces FPGA Accelerated Functions(AF).

Reply:
OK, let's see if I understand....

- First mode - you have the plugin in region mode, the webhook in orchestration mode, and the crio prestart hook is used to do the bitstream programming
- Second mode - you have the plugin in af mode, the webhook in preprogrammed mode, and the crio prestart hook is effectively un-used.

Or.... is the crio prestart hook always potentially used, in both 'modes' ?

Reply:
Yes, you understood it correctly. As second mode doesn't need a hook it means it doesn't require CRI-O and can use any other runtime (Docker, containerd, etc).

 As the hook is not needed in preprogrammed mode people should not use initcontainer in the plugin daemonset to avoid calling hook. If hook is called in preprogrammed mode it will fail and user workflow will not start.

Reply:
after discussion... I think we agreed the crio hook will not fail if used in `af`/`preprogrammed` mode, as the information it uses to program bitstreams will not be set/passed to it - so it should be benign.
As right now we do not have YAML examples without the crio hook initcontainer in them, I'll try to word appropriately. Basically, today, if you want to use `region`/`orchestration` mode, you will have to do some moderate modifications to the supplied YAML or deploy by hand.